### PR TITLE
Support code references at any level within extension's properties

### PR DESCRIPTION
**Pull request by vojtechszocs**
With this PR, `CodeRef<T>` values can appear at any level within extension's properties object.

For example, using the following Console extension declaration

```ts
export type Foo = ExtensionDeclaration<'console.foo', {
  bar: CodeRef<VoidFunction>;
  baz: { qux: CodeRef<VoidFunction> };
}>;
```

both `e.properties.bar` and `e.properties.baz.qux` values will be
1. parsed (`EncodedCodeRef` to `CodeRef`) when loading the given plugin
2. resolved (`CodeRef<T>` to `T`) when consumed via `useResolvedExtensions` hook

Detection of code references (both encoded & executable) relies on a recursive equivalent of [`_.forOwn`](https://docs-lodash.com/v4/for-own/) function that traverses plain objects and arrays. This means that e.g. `CodeRef<VoidFunction>[]` (array of code references) properties are supported as well.

cc @bipuladh @christianvogt

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/tsconfig.json
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/tsconfig.json:12]: Another alternative is to use existing `"target": "es2016"` with `"lib": ["es2016", "es2020.promise", "dom"]`

Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/tsconfig.json]: > Another alternative is to use `"target": "es2016"` with `"lib": ["es2016", "es2020.promise", "dom"]` to add specific types we need on top of the existing target spec.

I've updated the PR to use this approach.

**Comment by yapei** [comment]
Tested with following steps
```
in one terminal,
$ git clone git@github.com:vojtechszocs/console.git
$ cd console; git checkout nested-coderefs
$ git apply nested-coderefs-plugin-test.diff
$ cd frontend; yarn install
$ cd dynamic-demo-plugin; yarn install; yarn build; yarn http-server

in another terminal,
$ cd /back/to/console/root; ./build.sh
$ ./bin/bridge -plugins console-demo-plugin=http://localhost:9001/
```
then check browser console logs, we can see
```
Dynamic plugins: [console-demo-plugin]
Loading plugin manifest from http://localhost:9000/api/plugins/console-demo-plugin/plugin-manifest.json
Loading entry script for plugin console-demo-plugin@0.0.0 from http://localhost:9000/api/plugins/console-demo-plugin/plugin-entry.js
Added plugin console-demo-plugin@0.0.0
main-chunk-7d14ad1bc734ef26df43.min.js:1 Plugin console-demo-plugin@0.0.0 is now enabled
bar.ts:4 testHandler called (e,t)=>{f.b.dispatch(y(e,t))}
bar.ts:8 testing 1
bar.ts:8 testing 2
```

**Review by bipuladh — APPROVED** [review]
/lgtm

**Comment by bipuladh** [comment]
```
Unable to connect to the server: dial tcp: lookup api.ci-op-hk8n5pm2-75d12.********************************* on 172.30.0.10:53: no such host
```
Looks like the test-cluster is unable to come up.

**Comment by vojtechszocs** [comment]
/label px-approved
/label docs-approved

We have a general px- and docs-approval for Console dynamic plugin changes.

**Comment by vojtechszocs** [comment]
/hold

`e2e-gcp-console` is failing consistently, need to investigate.

**Comment by jerolimov** [comment]
Hey @vojtechszocs, I investigate why the E2E tests fails. They missed the guided tour when the user switches the first time to the developer perspective.

And thanks these tests I can confirm with some manual tests now that this alert / modal doesn't appear anymore. The guided tour is also not shown anymore in the dropdown menu in the top right corner.

Menu for kubeadmin before:

before-kubeadmin-menu

Menu for kubeadmin with this change:

after-kubeadmin-menu

Similar for a developer before:

before-developer-menu

Similar for a developer with this change:

after-developer-menu

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/__tests__/url.spec.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/__tests__/url.spec.ts:9]:
```suggestion
    expect(resolveURL('http://test/foo', 'bar', getDocumentOrigin)).toBe('http://test/foo/bar');
    expect(resolveURL('http://test/foo/', 'bar', getDocumentOrigin)).toBe('http://test/foo/bar');
```

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/__tests__/url.spec.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/utils/__tests__/url.spec.ts:14]: A second case would be useful here as well:

```suggestion
    expect(resolveURL('/foo', 'bar', getDocumentOrigin)).toBe('https://example:1234/foo/bar');
    expect(resolveURL('/foo/', 'bar', getDocumentOrigin)).toBe('https://example:1234/foo/bar');
```

File: frontend/packages/console-dynamic-plugin-sdk/src/utils/store.ts
Comment (jerolimov) [reply frontend/packages/console-dynamic-plugin-sdk/src/utils/store.ts:10]: @vojtechszocs This is the change which breaks the guided tour! See above.

**Comment by jerolimov** [comment]
The last comment above shows that this PR breaks the devconsole Guided Tour. As mentioned, the change from `Object.assign` to `_.merge` in `store.ts` breaks this. See https://github.com/openshift/console/pull/9070

/lgtm cancel

**Comment by vojtechszocs** [comment]
@jerolimov Thanks so much for looking into this! I'll rebase & update PR accordingly.

File: frontend/packages/console-plugin-sdk/src/utils/extension-i18n.ts
Comment (jerolimov) [inline frontend/packages/console-plugin-sdk/src/utils/extension-i18n.ts:6]: I don't like micro optimizations, but I also think its not necessary to create/compile/match a regex pattern here.

```suggestion
  return typeof value === 'string' && value.startsWith('%') && value.endsWith('%');
```

**Comment by jerolimov** [comment]
Thanks for the update. The Guided tour is back again. I also tested a code ref on a second level by changing the ContextProvider extension locally. Works fine. :+1:

/lgtm

**Review by jerolimov — APPROVED** [review]

File: frontend/packages/console-plugin-sdk/src/utils/extension-i18n.ts
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/utils/extension-i18n.ts:6]: Agreed, we can use `startsWith '%' and endsWith '%' and length > 2` condition instead of RegExp here.

**Comment by jerolimov** [comment]
> ✖  addFlow/create-from-devfile.feature
> ✖  addFlow/create-from-docker-file.feature
> ✖  addFlow/create-from-git.feature

/retest

**Comment by vojtechszocs** [comment]
Rebased, updated `isTranslatableString` function & added a unit test.

**Review by jerolimov — APPROVED** [review]
Tested nested code refs again and works. Guided tour also works.

/lgtm

**Pull request merged**
